presto-cache/src/main/java/com/facebook/presto/cache/LocalRangeCacheManager.java

```diff
@@ -0,0 +1,461 @@
+                Files.write((new File(newFilePath.toUri())).toPath(), previousFileBytes, CREATE_NEW);
+                int previousFileLength = previousFileBytes.length;
+                long previousFileOffset = previousCacheFile.getOffset();
+
+                // remove the overlapping part and append the remaining cache data
+                byte[] remainingCacheFileBytes = new byte[toIntExact((key.getLength() + key.getOffset()) - (previousFileLength + previousFileOffset))];
+                System.arraycopy(data, toIntExact(previousFileLength + previousFileOffset - key.getOffset()), remainingCacheFileBytes, 0, remainingCacheFileBytes.length);
+                Files.write((new File(newFilePath.toUri())).toPath(), remainingCacheFileBytes, APPEND);
+
+                // update new range info
+                newFileLength = previousFileLength + remainingCacheFileBytes.length;
+                newFileOffset = previousFileOffset;
+            }
+
+            if (followingCacheFile != null) {
+                // remove the overlapping part and append the remaining following file data
+                try (RandomAccessFile followingFile = new RandomAccessFile(new File(followingCacheFile.getPath().toUri()), "r")) {
+                    byte[] remainingFollowingFileBytes = new byte[toIntExact((followingFile.length() + followingCacheFile.getOffset()) - (key.getLength() + key.getOffset()))];
+                    followingFile.seek(key.getOffset() + key.getLength() - followingCacheFile.getOffset());
+                    followingFile.readFully(remainingFollowingFileBytes, 0, remainingFollowingFileBytes.length);
+                    Files.write((new File(newFilePath.toUri())).toPath(), remainingFollowingFileBytes, APPEND);
+
+                    // update new range info
+                    newFileLength += remainingFollowingFileBytes.length;
+                }
+            }
+        }
+        catch (IOException e) {
+            log.warn(e, "%s encountered an error while flushing file %s", Thread.currentThread().getName(), newFilePath);
+            tryDeleteFile(newFilePath);
+            return false;
+        }
+
+        boolean updated;
+        Set<Path> cacheFilesToDelete = new HashSet<>();
+        Lock writeLock = persisted.get(targetFile).getLock().writeLock();
+        writeLock.lock();
+        try {
+            RangeMap<Long, LocalCacheFile> cache = persisted.get(targetFile).getRange();
+            // check again if the previous or following range has been updated by someone else
+            LocalCacheFile newPreviousCacheFile = cache.get(key.getOffset() - 1);
+            LocalCacheFile newFollowingCacheFile = cache.get(key.getOffset() + key.getLength());
+
+            if (!cacheFileEquals(previousCacheFile, newPreviousCacheFile) || !cacheFileEquals(followingCacheFile, newFollowingCacheFile)) {
+                // someone else has updated the cache; delete the newly created file
+                updated = false;
```

Comment:
Let's add comment to state that this flag is used to avoid putting file deletion operations into write lock.